### PR TITLE
Forbid mutable class fields #891

### PR DESCRIPTION
**Pull request by PrVrSs**
# I have made things!

## Checklist

- [ ] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [ ] I have created at least one test case for the changes I have made
- [ ] I have updated the documentation for the changes I have made
- [ ] I have added my changes to the `CHANGELOG.md`

### CODE REVIEW

**Comment by PrVrSs** [comment]
what about defaultdict and etc?

**Comment by PrVrSs** [comment]
mb forbid only comprehension? https://twitter.com/doctaphred/status/1164935519974895622
dict/list we can use for example in __init_subclass__ and other class methods
what about Monostate ("Borg") http://www.aleax.it/gdd_pydp.pdf?
 Meta? https://www.usenix.org/system/files/login/articles/login_winter16_12_beazley.pdf
and use as a result
```
class Simple(Tokenizer):
 tokens = [
 (‘NAME’, r’[a-zA-Z_][a-zA-Z0-9_]*’),
 (‘NUM’, r’\d+’),
 (‘ASSIGN’, r’=’),
 (‘PLUS’, r’\+’),
 (‘TIMES’, ‘\*’),
 (‘ignore’, r’\s+’)
 ]
 actions = {
 ‘NAME’: lambda text: text.upper(),
 ‘NUM’: lambda text: int(text)
 }
```

File: tests/test_visitors/test_ast/test_classes/test_class_attributes/test_mutable_attributes.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    ClassMutableAttributeVisitor,
+)
+
+class_template = """
+class ClassWithAttr(object):
+    field = {0}
+"""
+
+mutable_assign = (
+    '[]',
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_classes/test_class_attributes/test_mutable_attributes.py]: We also need to test:
- `set()`, `{1, 2}`
- and all comprehensions

File: wemake_python_styleguide/violations/best_practices.py [hunk 2]
```diff
@@ -1855,3 +1857,38 @@ class StringConstantRedefinedViolation(ASTViolation):
...
+    Solution:
+        Use immutable types for class attributes.
+
+    Example::
+
+        # Correct:
+        class Test(object):
+            field = ('field1',)
+
+        class Test(object):
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py]: Just one correct is fine. And the same for the wrong example.

File: tests/test_visitors/test_ast/test_classes/test_class_attributes/test_mutable_attributes.py (hunk 1 above)
Comment (PrVrSs) [reply tests/test_visitors/test_ast/test_classes/test_class_attributes/test_mutable_attributes.py]: i didn't forbid `set`. what to do with `previous_codes`?

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_classes/test_class_attributes/test_mutable_attributes.py]: Oh, yaiks! 

They should really be `frozenset`s 💯

File: wemake_python_styleguide/violations/best_practices.py (hunk 2 above)
Comment (PrVrSs) [reply wemake_python_styleguide/violations/best_practices.py]: done

File: tests/test_visitors/test_ast/test_classes/test_class_attributes/test_mutable_attributes.py (hunk 1 above)
Comment (PrVrSs) [reply tests/test_visitors/test_ast/test_classes/test_class_attributes/test_mutable_attributes.py]: done

**Comment by sobolevn** [comment]
After a heated discussion with several python developers, I have decided to reconsider this feature.

There are valid cases to have class-level mutable fields.

@PrVrSs sorry for that. You have spent a significant amount of time and effort for useless work. That's due to my lack of planning. Please, send me your shipping address via email and I will send you a set of t-shirt and stickers as a compliment. 🙂

**Comment by PrVrSs** [comment]
sometimes it happens, np.

**Pull request closed without merging**